File: apps/rhino-cli-go/internal/envvalidate/extractor.go
package envvalidate

import (
	"sort"
	"strings"
	"unicode"
)

// ExtractRust extracts env var names from Rust source (env::var("KEY")).
func ExtractRust(content string) []string {
	return extractPattern(content, []string{`env::var("`}, "//")
}

// ExtractGo extracts env var names from Go source:
// os.Getenv("KEY"), os.LookupEnv("KEY"), struct tags `env:"KEY"`.
func ExtractGo(content string) []string {
	keys := make(map[string]struct{})
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "//") {
			continue
		}
		for _, pat := range []string{`os.Getenv("`, `os.LookupEnv("`} {
			rest := trimmed
			for {
				pos := strings.Index(rest, pat)
				if pos < 0 {
					break
				}
				after := rest[pos+len(pat):]
				end := strings.Index(after, `"`)
				if end >= 0 {
					k := after[:end]
					if isEnvKey(k) {
						keys[k] = struct{}{}
					}
				}
				rest = rest[pos+len(pat):]
			}
		}
		// struct tags: `env:"KEY"` or `env:"KEY,..."`
		rest := trimmed
		for {
			pos := strings.Index(rest, "`env:\"")
			if pos < 0 {
				break
			}
			after := rest[pos+6:]
			end := strings.Index(after, `"`)
			if end >= 0 {
				tagVal := after[:end]
				k := strings.SplitN(tagVal, ",", 2)[0]
				if isEnvKey(k) {
					keys[k] = struct{}{}
				}
			}
			rest = rest[pos+6:]
		}
	}
	return sortedKeys(keys)
}

// ExtractTypeScript extracts env var names from TypeScript/JavaScript source:
// process.env.KEY, process.env["KEY"], Config.string/integer/number/boolean("KEY").
func ExtractTypeScript(content string) []string {
	keys := make(map[string]struct{})
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "//") {
			continue
		}
		// process.env.KEY
		rest := trimmed
		for {
			pos := strings.Index(rest, "process.env.")
			if pos < 0 {
				break
			}
			after := rest[pos+12:]
			var sb strings.Builder
			for _, c := range after {
				if unicode.IsUpper(c) || c == '_' || unicode.IsDigit(c) {
					sb.WriteRune(c)
				} else {
					break
				}
			}
			k := sb.String()
			if isEnvKey(k) {
				keys[k] = struct{}{}
			}
			rest = rest[pos+12:]
		}
		// process.env["KEY"]
		rest = trimmed
		for {
			pos := strings.Index(rest, `process.env["`)
			if pos < 0 {
				break
			}
			after := rest[pos+13:]
			end := strings.Index(after, `"`)
			if end >= 0 {
				k := after[:end]
				if isEnvKey(k) {
					keys[k] = struct{}{}
				}
			}
			rest = rest[pos+13:]
		}
		// Config.string/integer/number/boolean("KEY")
		for _, variant := range []string{`Config.string("`, `Config.integer("`, `Config.number("`, `Config.boolean("`} {
			rest := trimmed
			for {
				pos := strings.Index(rest, variant)
				if pos < 0 {
					break
				}
				after := rest[pos+len(variant):]
				end := strings.Index(after, `"`)
				if end >= 0 {
					k := after[:end]
					if isEnvKey(k) {
						keys[k] = struct{}{}
					}
				}
				rest = rest[pos+len(variant):]
			}
		}
	}
	return sortedKeys(keys)
}

// ExtractClojure extracts env var names from Clojure source.
// Matches both (System/getenv "KEY") and (getenv "KEY") (local wrapper form).
func ExtractClojure(content string) []string {
	keys := make(map[string]struct{})
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, ";") {
			continue
		}
		rest := trimmed
		for {
			pos := strings.Index(rest, `getenv "`)
			if pos < 0 {
				break
			}
			after := rest[pos+8:]
			end := strings.Index(after, `"`)
			if end >= 0 {
				k := after[:end]
				if isEnvKey(k) {
					keys[k] = struct{}{}
				}
			}
			rest = rest[pos+8:]
		}
	}
	return sortedKeys(keys)
}

// ExtractCSharp extracts env var names from C# source:
// Configuration["KEY"], Environment.GetEnvironmentVariable("KEY").
func ExtractCSharp(content string) []string {
	return extractPattern(content, []string{`Configuration["`, `GetEnvironmentVariable("`}, "//")
}

// ExtractElixir extracts env var names from Elixir source (System.get_env("KEY")).
func ExtractElixir(content string) []string {
	return extractPattern(content, []string{`System.get_env("`}, "#")
}

// ExtractFSharp extracts env var names from F# source (Environment.GetEnvironmentVariable("KEY")).
func ExtractFSharp(content string) []string {
	return extractPattern(content, []string{`GetEnvironmentVariable("`}, "//")
}

// ExtractJava extracts env var names from Java source (System.getenv("KEY"))
// or Spring YAML (${KEY} without default).
func ExtractJava(content string, isYAML bool) []string {
	keys := make(map[string]struct{})
	if isYAML {
		for _, line := range strings.Split(content, "\n") {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "#") {
				continue
			}
			rest := trimmed
			for {
				pos := strings.Index(rest, "${")
				if pos < 0 {
					break
				}
				after := rest[pos+2:]
				hasDefault := false
				end := -1
				for i, c := range after {
					if c == ':' {
						hasDefault = true
						break
					}
					if c == '}' {
						end = i
						break
					}
				}
				if !hasDefault && end >= 0 {
					k := after[:end]
					if isEnvKey(k) {
						keys[k] = struct{}{}
					}
				}
				rest = rest[pos+2:]
			}
		}
	} else {
		for _, line := range strings.Split(content, "\n") {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "//") {
				continue
			}
			rest := trimmed
			for {
				pos := strings.Index(rest, `System.getenv("`)
				if pos < 0 {
					break
				}
				after := rest[pos+15:]
				end := strings.Index(after, `"`)
				if end >= 0 {
					k := after[:end]
					if isEnvKey(k) {
						keys[k] = struct{}{}
					}
				}
				rest = rest[pos+15:]
			}
		}
	}
	return sortedKeys(keys)
}

// ExtractKotlin extracts env var names from Kotlin source (System.getenv("KEY")).
func ExtractKotlin(content string) []string {
	return extractPattern(content, []string{`System.getenv("`}, "//")
}

// ExtractPython extracts env var names from Python pydantic-settings source
// and os.getenv/os.environ.get calls.
func ExtractPython(content string) []string {
	keys := make(map[string]struct{})
	inSettingsClass := false
	classIndent := -1

	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)

		// Detect class Settings(BaseSettings):
		if strings.HasPrefix(trimmed, "class ") && strings.Contains(trimmed, "BaseSettings") {
			inSettingsClass = true
			classIndent = -1
			continue
		}

		if !inSettingsClass {
			// Extract os.environ.get, os.environ[, os.getenv
			for _, pat := range []string{`os.environ.get("`, `os.environ["`, `os.getenv("`} {
				rest := trimmed
				for {
					pos := strings.Index(rest, pat)
					if pos < 0 {
						break
					}
					after := rest[pos+len(pat):]
					end := strings.Index(after, `"`)
					if end >= 0 {
						k := after[:end]
						if isEnvKey(k) {
							keys[k] = struct{}{}
						}
					}
					rest = rest[pos+len(pat):]
				}
			}
			continue
		}

		// Inside class body
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		indent := len(line) - len(strings.TrimLeft(line, " \t"))

		// Set class body indent on first non-empty line
		if classIndent < 0 && indent > 0 {
			classIndent = indent
		}

		// End of class: dedented past body indent
		if classIndent >= 0 && indent < classIndent {
			inSettingsClass = false
			continue
		}

		// Skip model_config lines
		if strings.HasPrefix(trimmed, "model_config") {
			continue
		}

		// Match field: identifier: type (with or without = default)
		colonPos := strings.Index(trimmed, ":")
		if colonPos > 0 {
			namePart := strings.TrimSpace(trimmed[:colonPos])
			if isValidPythonIdent(namePart) {
				k := strings.ToUpper(namePart)
				if isEnvKey(k) {
					keys[k] = struct{}{}
				}
			}
		}
	}
	return sortedKeys(keys)
}

// isValidPythonIdent returns true for a bare Python identifier (no spaces, starts with alpha/underscore).
func isValidPythonIdent(s string) bool {
	if s == "" {
		return false
	}
	for i, c := range s {
		if i == 0 {
			if !unicode.IsLetter(c) && c != '_' {
				return false
			}
		} else {
			if !unicode.IsLetter(c) && !unicode.IsDigit(c) && c != '_' {
				return false
			}
		}
	}
	return true
}

// extractPattern is a helper for extractors that share the same single-string-arg pattern.
// commentPrefix: skip lines starting with this prefix.
func extractPattern(content string, patterns []string, commentPrefix string) []string {
	keys := make(map[string]struct{})
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if commentPrefix != "" && strings.HasPrefix(trimmed, commentPrefix) {
			continue
		}
		for _, pat := range patterns {
			rest := trimmed
			for {
				pos := strings.Index(rest, pat)
				if pos < 0 {
					break
				}
				after := rest[pos+len(pat):]
				end := strings.Index(after, `"`)
				if end >= 0 {
					k := after[:end]
					if isEnvKey(k) {
						keys[k] = struct{}{}
					}
				}
				rest = rest[pos+len(pat):]
			}
		}
	}
	return sortedKeys(keys)
}

// isEnvKey returns true for a valid ALL_CAPS env var key (>=2 chars, starts with uppercase letter).
func isEnvKey(s string) bool {
	if len(s) < 2 {
		return false
	}
	for i, c := range s {
		if i == 0 {
			if !unicode.IsUpper(c) {
				return false
			}
		} else {
			if !unicode.IsUpper(c) && !unicode.IsDigit(c) && c != '_' {
				return false
			}
		}
	}
	return true
}

func sortedKeys(m map[string]struct{}) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: apps/rhino-cli-rust/src/internal/envvalidate/extractor.rs
//! Line-oriented env-read extractors, one per source language.
//!
//! All extractors use simple regex matches — no language parser dependency.
//! Known false-positive/negative modes are documented in tech-docs.md §6.1.

use std::collections::BTreeSet;

/// Extract env var names read by Rust source (`env::var("KEY")` / `std::env::var("KEY")`).
pub fn extract_rust(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("//") {
            continue;
        }
        let mut rest = line;
        while let Some(pos) = rest.find("env::var(\"") {
            rest = &rest[pos + 10..];
            if let Some(end) = rest.find('"') {
                let key = &rest[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
                rest = &rest[end..];
            } else {
                break;
            }
        }
    }
    keys
}

/// Extract env var names read by Go source:
/// - `os.Getenv("KEY")` / `os.LookupEnv("KEY")`
/// - struct tags `` `env:"KEY"` `` / `` `env:"KEY,required"` ``
pub fn extract_go(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("//") {
            continue;
        }
        // os.Getenv / os.LookupEnv
        for pattern in &[r"os.Getenv(", r"os.LookupEnv("] {
            let mut rest = line;
            while let Some(pos) = rest.find(pattern) {
                let after = &rest[pos + pattern.len()..];
                if let Some(after) = after.strip_prefix('"')
                    && let Some(end) = after.find('"')
                {
                    let key = &after[..end];
                    if is_env_key(key) {
                        keys.insert(key.to_string());
                    }
                }
                rest = &rest[pos + pattern.len()..];
            }
        }
        // struct tags: `env:"KEY"` or `env:"KEY,..."``
        let mut rest = line;
        while let Some(pos) = rest.find(r#"`env:""#) {
            let after = &rest[pos + 6..];
            if let Some(end) = after.find('"') {
                let tag_val = &after[..end];
                // Strip options like ",required"
                let key = tag_val.split(',').next().unwrap_or(tag_val);
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 6..];
        }
    }
    keys
}

/// Extract env var names read by TypeScript/JavaScript source:
/// - `process.env.KEY`
/// - `process.env["KEY"]`
/// - `Config.string("KEY")` / `Config.integer("KEY")` / `Config.number("KEY")` / `Config.boolean("KEY")`
pub fn extract_typescript(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("//") {
            continue;
        }
        // process.env.KEY
        let mut rest = line;
        while let Some(pos) = rest.find("process.env.") {
            let after = &rest[pos + 12..];
            let key: String = after
                .chars()
                .take_while(|c| c.is_ascii_uppercase() || *c == '_' || c.is_ascii_digit())
                .collect();
            if is_env_key(&key) {
                keys.insert(key);
            }
            rest = &rest[pos + 12..];
        }
        // process.env["KEY"]
        let mut rest = line;
        while let Some(pos) = rest.find("process.env[\"") {
            let after = &rest[pos + 13..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 13..];
        }
        // Config.string("KEY") / Config.integer("KEY") etc.
        for variant in &[
            "Config.string(\"",
            "Config.integer(\"",
            "Config.number(\"",
            "Config.boolean(\"",
        ] {
            let mut rest = line;
            while let Some(pos) = rest.find(variant) {
                let after = &rest[pos + variant.len()..];
                if let Some(end) = after.find('"') {
                    let key = &after[..end];
                    if is_env_key(key) {
                        keys.insert(key.to_string());
                    }
                }
                rest = &rest[pos + variant.len()..];
            }
        }
    }
    keys
}

/// Extract env var names read by Clojure source.
///
/// Matches both `(System/getenv "KEY")` (Java interop) and `(getenv "KEY")`
/// (local wrapper — the pedestal app defines `getenv` as a thin `System/getenv` shim).
pub fn extract_clojure(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with(';') {
            continue;
        }
        // Match `getenv "KEY"` — covers both `(System/getenv "KEY")` and `(getenv "KEY")`
        let mut rest = line;
        while let Some(pos) = rest.find("getenv \"") {
            let after = &rest[pos + 8..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 8..];
        }
    }
    keys
}

/// Extract env var names read by C# source:
/// - `Configuration["KEY"]`
/// - `Environment.GetEnvironmentVariable("KEY")`
pub fn extract_csharp(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("//") {
            continue;
        }
        // Configuration["KEY"]
        let mut rest = line;
        while let Some(pos) = rest.find("Configuration[\"") {
            let after = &rest[pos + 15..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 15..];
        }
        // GetEnvironmentVariable("KEY")
        let mut rest = line;
        while let Some(pos) = rest.find("GetEnvironmentVariable(\"") {
            let after = &rest[pos + 24..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 24..];
        }
    }
    keys
}

/// Extract env var names read by Elixir/Elixir config source: `System.get_env("KEY")`.
pub fn extract_elixir(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with('#') {
            continue;
        }
        let mut rest = line;
        while let Some(pos) = rest.find("System.get_env(\"") {
            let after = &rest[pos + 16..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 16..];
        }
    }
    keys
}

/// Extract env var names read by F# source: `Environment.GetEnvironmentVariable("KEY")`.
pub fn extract_fsharp(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("//") {
            continue;
        }
        let mut rest = line;
        while let Some(pos) = rest.find("GetEnvironmentVariable(\"") {
            let after = &rest[pos + 24..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 24..];
        }
    }
    keys
}

/// Extract env var names from Java source (`System.getenv("KEY")`) and Spring
/// YAML (`${KEY}` without default — i.e. `${KEY}` not `${KEY:default}`).
pub fn extract_java(content: &str, is_yaml: bool) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    if is_yaml {
        // YAML: ${KEY} (no colon = required; ${KEY:default} = optional, skip)
        for line in content.lines() {
            let line = line.trim();
            if line.starts_with('#') {
                continue;
            }
            let mut rest = line;
            while let Some(pos) = rest.find("${") {
                let after = &rest[pos + 2..];
                // Find end: '}' but stop if we see ':' first (that's a default)
                let mut end = None;
                let mut has_default = false;
                for (i, c) in after.char_indices() {
                    if c == ':' {
                        has_default = true;
                        break;
                    }
                    if c == '}' {
                        end = Some(i);
                        break;
                    }
                }
                if !has_default && let Some(end) = end {
                    let key = &after[..end];
                    if is_env_key(key) {
                        keys.insert(key.to_string());
                    }
                }
                rest = &rest[pos + 2..];
            }
        }
    } else {
        // Java: System.getenv("KEY")
        for line in content.lines() {
            let line = line.trim();
            if line.starts_with("//") {
                continue;
            }
            let mut rest = line;
            while let Some(pos) = rest.find("System.getenv(\"") {
                let after = &rest[pos + 15..];
                if let Some(end) = after.find('"') {
                    let key = &after[..end];
                    if is_env_key(key) {
                        keys.insert(key.to_string());
                    }
                }
                rest = &rest[pos + 15..];
            }
        }
    }
    keys
}

/// Extract env var names read by Kotlin source: `System.getenv("KEY")`.
pub fn extract_kotlin(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    for line in content.lines() {
        let line = line.trim();
        if line.starts_with("//") {
            continue;
        }
        let mut rest = line;
        while let Some(pos) = rest.find("System.getenv(\"") {
            let after = &rest[pos + 15..];
            if let Some(end) = after.find('"') {
                let key = &after[..end];
                if is_env_key(key) {
                    keys.insert(key.to_string());
                }
            }
            rest = &rest[pos + 15..];
        }
    }
    keys
}

/// Extract env var names from Python pydantic-settings source.
///
/// Scans for field declarations inside `class Settings(BaseSettings)`:
/// - Required fields: `fieldname: type` (no `= default`)
/// - Optional fields: `fieldname: type = default` (also extracted; optional ones go in allowlist)
///
/// Field names are upper-cased to match env var convention.
pub fn extract_python(content: &str) -> BTreeSet<String> {
    let mut keys = BTreeSet::new();
    let mut in_settings_class = false;
    let mut class_indent: Option<usize> = None;

    for line in content.lines() {
        let trimmed = line.trim();

        // Detect `class Settings(BaseSettings):` (or subclass)
        if trimmed.starts_with("class ") && trimmed.contains("BaseSettings") {
            in_settings_class = true;
            class_indent = None;
            continue;
        }

        if !in_settings_class {
            // Also extract os.environ.get and os.getenv
            for pat in &["os.environ.get(\"", "os.environ[\"", "os.getenv(\""] {
                let mut rest = trimmed;
                while let Some(pos) = rest.find(pat) {
                    let after = &rest[pos + pat.len()..];
                    if let Some(end) = after.find('"') {
                        let key = &after[..end];
                        if is_env_key(key) {
                            keys.insert(key.to_string());
                        }
                    }
                    rest = &rest[pos + pat.len()..];
                }
            }
            continue;
        }

        // Inside class body
        let indent = line.len() - line.trim_start().len();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        // Detect end of class (line dedented past class body indent)
        if let Some(ci) = class_indent
            && indent < ci
        {
            in_settings_class = false;
            continue;
        }

        // Set class body indent on first non-empty line
        if class_indent.is_none() && indent > 0 {
            class_indent = Some(indent);
        }

        // Skip model_config lines
        if trimmed.starts_with("model_config") {
            continue;
        }

        // Match field: `identifier: type` (with or without `= default`)
        // Identifier must be a valid Python name starting with letter/underscore.
        if let Some(colon_pos) = trimmed.find(':') {
            let name_part = &trimmed[..colon_pos];
            // Must be a pure identifier (no spaces, starts with alpha/underscore)
            let name = name_part.trim();
            if !name.is_empty()
                && name
                    .chars()
                    .next()
                    .is_some_and(|c| c.is_alphabetic() || c == '_')
                && name.chars().all(|c| c.is_alphanumeric() || c == '_')
            {
                // Convert snake_case to UPPER_CASE for env var name
                let key = name.to_uppercase();
                if is_env_key(&key) {
                    keys.insert(key);
                }
            }
        }
    }
    keys
}

/// Returns true for a valid ALL_CAPS env var name (at least 2 chars, starts with letter).
fn is_env_key(s: &str) -> bool {
    if s.len() < 2 {
        return false;
    }
    let mut chars = s.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    if !first.is_ascii_uppercase() {
        return false;
    }
    chars.all(|c| c.is_ascii_uppercase() || c.is_ascii_digit() || c == '_')
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use super::*;

    fn set(keys: &[&str]) -> BTreeSet<String> {
        keys.iter().map(std::string::ToString::to_string).collect()
    }

    #[test]
    fn rust_extracts_env_var_calls() {
        let src = r#"
            let secret = env::var("FIXTURE_JWT_SECRET").context("missing")?;
            let port = env::var("FIXTURE_PORT").unwrap_or("8080".into());
            // let comment = env::var("COMMENTED_OUT");
        "#;
        let got = extract_rust(src);
        assert_eq!(got, set(&["FIXTURE_JWT_SECRET", "FIXTURE_PORT"]));
    }

    #[test]
    fn rust_empty_on_no_matches() {
        let got = extract_rust("fn main() { println!(\"hello\"); }");
        assert!(got.is_empty());
    }

    #[test]
    fn go_extracts_getenv_and_struct_tags() {
        let src = r#"
            os.Getenv("FIXTURE_JWT_SECRET")
            JWTSecret string `env:"CRUD_BE_GOLANG_GIN_JWT_SECRET,required"`
            Port      string `env:"CRUD_BE_GOLANG_GIN_PORT" envDefault:"8201"`
        "#;
        let got = extract_go(src);
        assert!(got.contains("FIXTURE_JWT_SECRET"));
        assert!(got.contains("CRUD_BE_GOLANG_GIN_JWT_SECRET"));
        assert!(got.contains("CRUD_BE_GOLANG_GIN_PORT"));
    }

    #[test]
    fn typescript_extracts_process_env_and_config() {
        let src = r#"
            const url = process.env.BACKEND_URL;
            const secret = process.env["CRUD_FS_TS_NEXTJS_JWT_SECRET"];
            const jwt = yield* Config.string("CRUD_BE_TS_EFFECT_JWT_SECRET");
            const port = yield* Config.integer("CRUD_BE_TS_EFFECT_PORT");
        "#;
        let got = extract_typescript(src);
        assert!(got.contains("BACKEND_URL"));
        assert!(got.contains("CRUD_FS_TS_NEXTJS_JWT_SECRET"));
        assert!(got.contains("CRUD_BE_TS_EFFECT_JWT_SECRET"));
        assert!(got.contains("CRUD_BE_TS_EFFECT_PORT"));
    }

    #[test]
    fn clojure_extracts_system_getenv() {
        let src = r#"
            (getenv "CRUD_BE_CLOJURE_PEDESTAL_JWT_SECRET")
            ; (System/getenv "COMMENTED")
        "#;
        let got = extract_clojure(src);
        assert!(got.contains("CRUD_BE_CLOJURE_PEDESTAL_JWT_SECRET"));
        assert!(!got.contains("COMMENTED"));
    }

    #[test]
    fn csharp_extracts_configuration_and_getenv() {
        let src = r#"
            builder.Configuration["CRUD_BE_CSHARP_ASPNETCORE_JWT_SECRET"]
            Environment.GetEnvironmentVariable("DATABASE_URL")
        "#;
        let got = extract_csharp(src);
        assert!(got.contains("CRUD_BE_CSHARP_ASPNETCORE_JWT_SECRET"));
        assert!(got.contains("DATABASE_URL"));
    }

    #[test]
    fn elixir_extracts_system_get_env() {
        let src = r#"
            System.get_env("CRUD_BE_ELIXIR_PHOENIX_JWT_SECRET")
            # System.get_env("COMMENTED")
        "#;
        let got = extract_elixir(src);
        assert!(got.contains("CRUD_BE_ELIXIR_PHOENIX_JWT_SECRET"));
        assert!(!got.contains("COMMENTED"));
    }

    #[test]
    fn fsharp_extracts_get_environment_variable() {
        let src = r#"
            Environment.GetEnvironmentVariable("CRUD_BE_FSHARP_GIRAFFE_JWT_SECRET")
        "#;
        let got = extract_fsharp(src);
        assert!(got.contains("CRUD_BE_FSHARP_GIRAFFE_JWT_SECRET"));
    }

    #[test]
    fn java_yaml_extracts_required_placeholder() {
        let src = r"
            secret: ${CRUD_BE_JAVA_SPRINGBOOT_JWT_SECRET}
            enabled: ${ENABLE_TEST_API:false}
        ";
        let got = extract_java(src, true);
        assert!(got.contains("CRUD_BE_JAVA_SPRINGBOOT_JWT_SECRET"));
        // ${ENABLE_TEST_API:false} has a default — should be omitted
        assert!(!got.contains("ENABLE_TEST_API"));
    }

    #[test]
    fn java_source_extracts_system_getenv() {
        let src = r#"
            String secret = System.getenv("CRUD_BE_JAVA_VERTX_JWT_SECRET");
        "#;
        let got = extract_java(src, false);
        assert!(got.contains("CRUD_BE_JAVA_VERTX_JWT_SECRET"));
    }

    #[test]
    fn kotlin_extracts_system_getenv() {
        let src = r#"
            val secret = System.getenv("CRUD_BE_KOTLIN_KTOR_JWT_SECRET")
        "#;
        let got = extract_kotlin(src);
        assert!(got.contains("CRUD_BE_KOTLIN_KTOR_JWT_SECRET"));
    }

    #[test]
    fn python_extracts_pydantic_fields_and_os_getenv() {
        let src = r#"
class Settings(BaseSettings):
    model_config = SettingsConfigDict(env_file=".env")
    database_url: str = "sqlite:///:memory:"
    crud_be_python_fastapi_jwt_secret: str
    app_jwt_issuer: str = "crud-be"

settings = Settings()
"#;
        let got = extract_python(src);
        assert!(got.contains("DATABASE_URL"));
        assert!(got.contains("CRUD_BE_PYTHON_FASTAPI_JWT_SECRET"));
        assert!(got.contains("APP_JWT_ISSUER"));
    }

    #[test]
    fn python_extracts_os_getenv_outside_class() {
        let src = r#"
import os
x = os.getenv("ENABLE_TEST_API")
y = os.environ.get("DATABASE_URL", "")
"#;
        let got = extract_python(src);
        assert!(got.contains("ENABLE_TEST_API"));
        assert!(got.contains("DATABASE_URL"));
    }

    #[test]
    fn is_env_key_filters_short_or_lowercase() {
        assert!(is_env_key("FOO"));
        assert!(is_env_key("CRUD_BE_RUST_AXUM_JWT_SECRET"));
        assert!(!is_env_key("foo"));
        assert!(!is_env_key("F")); // too short
        assert!(!is_env_key(""));
    }
}
